Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,750 @@
+        failureType: Failure.Type = Failure.self,
+        isWritable: Bool,
+        delegate: Delegate
+    ) -> NewWriter {
+        let writer = Self(
+            isWritable: isWritable,
+            delegate: delegate
+        )
+        let sink = Sink(storage: writer.storage)
+
+        return .init(sink: sink, writer: writer)
+    }
+
+    private init(
+        isWritable: Bool,
+        delegate: Delegate
+    ) {
+        let storage = Storage(
+            isWritable: isWritable,
+            delegate: delegate
+        )
+        self.internalClass = .init(storage: storage)
+    }
+
+    /// Yields a sequence of new elements to the ``NIOAsyncWriter``.
+    ///
+    /// If the ``NIOAsyncWriter`` is writable the sequence will get forwarded to the ``NIOAsyncWriterDelegate`` immediately.
+    /// Otherwise, the call to ``NIOAsyncWriter/yield(contentsOf:)`` will get suspended until the ``NIOAsyncWriter``
+    /// becomes writable again.
+    ///
+    /// If the ``NIOAsyncWriter`` is finished while a call to ``NIOAsyncWriter/yield(contentsOf:)`` is suspended the
+    /// call will throw a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// This can be called more than once and from multiple `Task`s at the same time.
+    ///
+    /// - Parameter contentsOf: The sequence to yield.
+    @inlinable
+    public func yield<S: Sequence>(contentsOf sequence: S) async throws where S.Element == Element {
+        try await self.storage.yield(contentsOf: sequence)
+    }
+
+    /// Yields an element to the ``NIOAsyncWriter``.
+    ///
+    /// If the ``NIOAsyncWriter`` is writable the element will get forwarded to the ``NIOAsyncWriterDelegate`` immediately.
+    /// Otherwise, the call to ``NIOAsyncWriter/yield(contentsOf:)`` will get suspended until the ``NIOAsyncWriter``
+    /// becomes writable again.
+    ///
+    /// If the ``NIOAsyncWriter`` is finished while a call to ``NIOAsyncWriter/yield(contentsOf:)`` is suspended the
+    /// call will throw a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// This can be called more than once and from multiple `Task`s at the same time.
+    ///
+    /// - Parameter element: The element to yield.
+    @inlinable
+    public func yield(_ element: Element) async throws {
+        try await self.yield(contentsOf: CollectionOfOne(element))
+    }
+
+    /// Finishes the writer.
+    ///
+    /// Calling this function signals the writer that any suspended or subsequent calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will return a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// - Note: Calling this function more than once has no effect.
+    /// - Parameter completion: A ``NIOAsyncWriterCompletion`` which indicates whether publishing has finished normally
+    /// or failed with an error.
+    @inlinable
+    public func finish(completion: NIOAsyncWriterCompletion<Failure>) {
```

Comment:
Is it actually worth having a _public_ `NIOAsyncWriterCompletion` here rather than say, `finish()` and `finish(error:)`?

`finish(completion: .finished)` and `finish(completion: .failure(.someError))` seem unnecessarily verbose.

Reply:
I adopted the spelling from `Combine`. We also have the spelling from `AsyncThrowingStream.Continuation` which has `finish(throwing error: Failure? = nil`.
I like your approach the best though having to separate methods